In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import gc                         
gc.enable() 

From the EDA, we get the solution reorder is highly related to:
1. customer behavior 
2. product characteristics
3. customer&product relationship

What's the next? We want to predict which previously purchased products (prior orders) will be in a user’s next order (train and test orders).

# Data preparation

In [51]:
aisles = pd.read_csv('aisles.csv')
departments = pd.read_csv('departments.csv')
order_products_prior = pd.read_csv('order_products__prior.csv')
order_products_train = pd.read_csv('order_products__train.csv')
orders = pd.read_csv('orders.csv')
products = pd.read_csv('products.csv')
sample_submission = pd.read_csv('sample_submission.csv')

In [52]:
# We convert character variables into category. 
# In Python, a categorical variable is called category and has a fixed number of different values
aisles['aisle'] = aisles['aisle'].astype('category')
departments['department'] = departments['department'].astype('category')
orders['eval_set'] = orders['eval_set'].astype('category')
products['product_name'] = products['product_name'].astype('category')

In [53]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                category
 3   order_number            int64   
 4   order_dow               int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
dtypes: category(1), float64(1), int64(5)
memory usage: 159.9 MB


In [54]:
order_products_prior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB


In [61]:
order_products_prior.head(20)

order_id  product_id  add_to_cart_order  reordered
0          2       33120                  1          1
1          2       28985                  2          1
2          2        9327                  3          0
3          2       45918                  4          1
4          2       30035                  5          0
5          2       17794                  6          1
6          2       40141                  7          1
7          2        1819                  8          1
8          2       43668                  9          0
9          3       33754                  1          1
10         3       24838                  2          1
11         3       17704                  3          1
12         3       21903                  4          1
13         3       17668                  5          1
14         3       46667                  6          1
15         3       17461                  7          1
16         3       32665                  8          1
17         4       46842                  1          0
18         4       26434                  2          1
19         4       39758                  3          1

In [60]:
order_products_train.head(20)

order_id  product_id  add_to_cart_order  reordered
0          1       49302                  1          1
1          1       11109                  2          1
2          1       10246                  3          0
3          1       49683                  4          0
4          1       43633                  5          1
5          1       13176                  6          0
6          1       47209                  7          0
7          1       22035                  8          1
8         36       39612                  1          0
9         36       19660                  2          1
10        36       49235                  3          0
11        36       43086                  4          1
12        36       46620                  5          1
13        36       34497                  6          1
14        36       48679                  7          1
15        36       46979                  8          1
16        38       11913                  1          0
17        38       18159                  2          0
18        38        4461                  3          0
19        38       21616                  4          1

In [57]:
# combine orders with
op = orders.merge(order_products_prior, on='order_id', how='inner')
op.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

In [58]:
op['eval_set'].value_counts()

prior    32434489
train           0
test            0
Name: eval_set, dtype: int64

# Feature Engineering

From the EDA, we clearly know re-order or not is highly related to:
1. customer behavior(eg: number of items per order)
2. product characteristics(eg: product category)
3. customer - product relationship (reorder times for a specific product)

Based on what we explored, we'll use following as predictors:
1. Numbers of orders per customer
2. How frequent a cutomer has reordered products
3. Number of purchases for each product
4. What's the probability for a product to be reordered
5. How many times a user bought a product
6. How frequently a customer bought a product after its first purchase
7. How many times a customer bought a product on its last 5 orders

In [9]:
#1 user numbers
user = op.groupby('user_id')['order_number'].max().to_frame('u_total_orders')
user = user.reset_index()
#user.head()

In [10]:
user.head()

user_id  u_total_orders
0        1              10
1        2              14
2        3              12
3        4               5
4        5               4

In [11]:
#2 frequency(how many times the user reordered)
user_reorder = op.groupby('user_id')['reordered'].mean().to_frame('user_reordered_ratio')
user_reorder = user_reorder.reset_index()
user_reorder.head()

user_id  user_reordered_ratio
0        1              0.694915
1        2              0.476923
2        3              0.625000
3        4              0.055556
4        5              0.378378

In [12]:
#3 purchase per order
order_purchase = op.groupby('product_id')['order_id'].count().to_frame('purchase_per_order')
order_purchase.reset_index()
order_purchase.head()

purchase_per_order
product_id                    
1                         1852
2                           90
3                          277
4                          329
5                           15

In [13]:
# 4 probability for a product to be reordered
## Remove products with less than 40 purchases
p_reorder = op.groupby('product_id').filter(lambda x: x.shape[0] >40)
p_reorder.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

In [14]:
##more than 40 purchase products P
p_reorder = p_reorder.groupby('product_id')['reordered'].mean().to_frame('p_reorder_ratio')
p_reorder = p_reorder.reset_index()
p_reorder.head()

product_id  p_reorder_ratio
0           1         0.613391
1           2         0.133333
2           3         0.732852
3           4         0.446809
4           8         0.503030

In [15]:
#Merge the order_product DataFrame with reorder
order_purchase = order_purchase.merge(p_reorder, on='product_id', how='left')

#delete the reorder DataFrame
del p_reorder
gc.collect()

order_purchase.head()

product_id  purchase_per_order  p_reorder_ratio
0           1                1852         0.613391
1           2                  90         0.133333
2           3                 277         0.732852
3           4                 329         0.446809
4           5                  15              NaN

In [16]:
#fill null value with 0
order_purchase['p_reorder_ratio'] = order_purchase['p_reorder_ratio'].fillna(value=0)
order_purchase.head()

product_id  purchase_per_order  p_reorder_ratio
0           1                1852         0.613391
1           2                  90         0.133333
2           3                 277         0.732852
3           4                 329         0.446809
4           5                  15         0.000000

In [17]:
user = user.merge(user_reorder, on='user_id', how='left')

del user_reorder
gc.collect()

user.head()

user_id  u_total_orders  user_reordered_ratio
0        1              10              0.694915
1        2              14              0.476923
2        3              12              0.625000
3        4               5              0.055556
4        5               4              0.378378

In [18]:
# 5 how many times a user bought a product
# Create distinct groups for each combination of user and product, count orders, save the result for each user X product to a new DataFrame 
uxp = op.groupby(['user_id', 'product_id'])['order_id'].count().to_frame('uxp_total_bought')
uxp = uxp.reset_index()
uxp.head()

user_id  product_id  uxp_total_bought
0        1         196                10
1        1       10258                 9
2        1       10326                 1
3        1       12427                10
4        1       13032                 3

In [19]:
# 6 How frequently a customer bought a product after its first purchase
# numerator
times = op.groupby(['user_id', 'product_id'])[['order_id']].count()
times.columns = ['Times_Bought_N']
times.head()


Times_Bought_N
user_id product_id                
1       196                     10
        10258                    9
        10326                    1
        12427                   10
        13032                    3

In [20]:
# denumerator
total_orders = op.groupby('user_id')['order_number'].max().to_frame('total_orders')
total_orders.head()

total_orders
user_id              
1                  10
2                  14
3                  12
4                   5
5                   4

In [21]:
first_order_no = op.groupby(['user_id', 'product_id'])['order_number'].min().to_frame('first_order_number')
first_order_no  = first_order_no.reset_index()
first_order_no.head()

user_id  product_id  first_order_number
0        1         196                   1
1        1       10258                   2
2        1       10326                   5
3        1       12427                   1
4        1       13032                   2

In [22]:
span = pd.merge(total_orders, first_order_no, on='user_id', how='right')
span.head()

user_id  total_orders  product_id  first_order_number
0        1            10         196                   1
1        1            10       10258                   2
2        1            10       10326                   5
3        1            10       12427                   1
4        1            10       13032                   2

In [23]:
span['Order_Range_D'] = span.total_orders - span.first_order_number + 1
span.head()

user_id  total_orders  product_id  first_order_number  Order_Range_D
0        1            10         196                   1             10
1        1            10       10258                   2              9
2        1            10       10326                   5              6
3        1            10       12427                   1             10
4        1            10       13032                   2              9

In [24]:
uxp_ratio = pd.merge(times, span, on=['user_id', 'product_id'], how='left')
uxp_ratio.head()

user_id  product_id  Times_Bought_N  total_orders  first_order_number  \
0        1         196              10            10                   1   
1        1       10258               9            10                   2   
2        1       10326               1            10                   5   
3        1       12427              10            10                   1   
4        1       13032               3            10                   2   

   Order_Range_D  
0             10  
1              9  
2              6  
3             10  
4              9

In [25]:
uxp_ratio['uxp_reorder_ratio'] = uxp_ratio.Times_Bought_N / uxp_ratio['Order_Range_D']
uxp_ratio.head()

user_id  product_id  Times_Bought_N  total_orders  first_order_number  \
0        1         196              10            10                   1   
1        1       10258               9            10                   2   
2        1       10326               1            10                   5   
3        1       12427              10            10                   1   
4        1       13032               3            10                   2   

   Order_Range_D  uxp_reorder_ratio  
0             10           1.000000  
1              9           1.000000  
2              6           0.166667  
3             10           1.000000  
4              9           0.333333

In [26]:
uxp_ratio = uxp_ratio.drop(['Times_Bought_N', 'total_orders', 'first_order_number', 'Order_Range_D'], axis=1)
uxp_ratio.head()

user_id  product_id  uxp_reorder_ratio
0        1         196           1.000000
1        1       10258           1.000000
2        1       10326           0.166667
3        1       12427           1.000000
4        1       13032           0.333333

In [27]:
uxp = uxp.merge(uxp_ratio, on=['user_id', 'product_id'], how='left')

del uxp_ratio
uxp.head()

user_id  product_id  uxp_total_bought  uxp_reorder_ratio
0        1         196                10           1.000000
1        1       10258                 9           1.000000
2        1       10326                 1           0.166667
3        1       12427                10           1.000000
4        1       13032                 3           0.333333

In [28]:
# 7 How many times a customer bought a product on its last 5 orders
op['order_number_back'] = op.groupby('user_id')['order_number'].transform(max) - op.order_number +1 
op5 = op[op.order_number_back <= 5]
last_five = op5.groupby(['user_id','product_id'])[['order_id']].count()
last_five.columns = ['times_last5']
last_five.head(10)

times_last5
user_id product_id             
1       196                   5
        10258                 5
        12427                 5
        13032                 2
        25133                 5
        35951                 1
        38928                 1
        39657                 1
        46149                 3
        49235                 2

Eg: So for user_id==1, the product 196 has been ordered on all of its last five orders, where the product 35951 has been ordered only one time.

In [29]:
# Merge the final feature with uxp DataFrame
uxp = uxp.merge(last_five, on=['user_id', 'product_id'], how='left')

del [op5 , last_five]
uxp.head()

user_id  product_id  uxp_total_bought  uxp_reorder_ratio  times_last5
0        1         196                10           1.000000          5.0
1        1       10258                 9           1.000000          5.0
2        1       10326                 1           0.166667          NaN
3        1       12427                10           1.000000          5.0
4        1       13032                 3           0.333333          2.0

In [30]:
# fillna
uxp = uxp.fillna(0)
uxp.head()

user_id  product_id  uxp_total_bought  uxp_reorder_ratio  times_last5
0        1         196                10           1.000000          5.0
1        1       10258                 9           1.000000          5.0
2        1       10326                 1           0.166667          0.0
3        1       12427                10           1.000000          5.0
4        1       13032                 3           0.333333          2.0

In [31]:
# Merge all feature together
data = uxp.merge(user, on='user_id', how='left')
data.head()

user_id  product_id  uxp_total_bought  uxp_reorder_ratio  times_last5  \
0        1         196                10           1.000000          5.0   
1        1       10258                 9           1.000000          5.0   
2        1       10326                 1           0.166667          0.0   
3        1       12427                10           1.000000          5.0   
4        1       13032                 3           0.333333          2.0   

   u_total_orders  user_reordered_ratio  
0              10              0.694915  
1              10              0.694915  
2              10              0.694915  
3              10              0.694915  
4              10              0.694915

# Train& Test Split

In [32]:
orders_future = orders[((orders.eval_set=='train') | (orders.eval_set=='test'))] #we only need train or test, no need for prior
orders_future = orders_future[ ['user_id', 'eval_set', 'order_id'] ]
orders_future.head(10)

user_id eval_set  order_id
10        1    train   1187899
25        2    train   1492625
38        3     test   2774568
44        4     test    329954
49        5    train   2196797
53        6     test   1528013
74        7    train    525192
78        8    train    880375
82        9    train   1094988
88       10    train   1822501

In [36]:
data = data.merge(orders_future, on='user_id', how='left')
data.head(15)

user_id  product_id  uxp_total_bought  uxp_reorder_ratio  times_last5  \
0         1         196                10           1.000000          5.0   
1         1       10258                 9           1.000000          5.0   
2         1       10326                 1           0.166667          0.0   
3         1       12427                10           1.000000          5.0   
4         1       13032                 3           0.333333          2.0   
5         1       13176                 2           0.222222          0.0   
6         1       14084                 1           0.100000          0.0   
7         1       17122                 1           0.166667          0.0   
8         1       25133                 8           1.000000          5.0   
9         1       26088                 2           0.200000          0.0   
10        1       26405                 2           0.200000          0.0   
11        1       30450                 1           0.125000          0.0   
12        1       35951                 1           1.000000          1.0   
13        1       38928                 1           1.000000          1.0   
14        1       39657                 1           1.000000          1.0   

    u_total_orders  user_reordered_ratio eval_set_x  order_id_x eval_set_y  \
0               10              0.694915      train     1187899      train   
1               10              0.694915      train     1187899      train   
2               10              0.694915      train     1187899      train   
3               10              0.694915      train     1187899      train   
4               10              0.694915      train     1187899      train   
5               10              0.694915      train     1187899      train   
6               10              0.694915      train     1187899      train   
7               10              0.694915      train     1187899      train   
8               10              0.694915      train     1187899      train   
9               10              0.694915      train     1187899      train   
10              10              0.694915      train     1187899      train   
11              10              0.694915      train     1187899      train   
12              10              0.694915      train     1187899      train   
13              10              0.694915      train     1187899      train   
14              10              0.694915      train     1187899      train   

    order_id_y  
0      1187899  
1      1187899  
2      1187899  
3      1187899  
4      1187899  
5      1187899  
6      1187899  
7      1187899  
8      1187899  
9      1187899  
10     1187899  
11     1187899  
12     1187899  
13     1187899  
14     1187899